src/persist-client/src/stats.rs

```diff
@@ -137,30 +137,29 @@ where
     let records = updates.records();
 
     if format.is_structured() {
-        // At the moment, the only way to collect stats is to re-encode the key column.
-        // However, we only need to re-encode the structured values if they weren't passed in.
-        // TODO(bkirwi): reuse the existing encoded data once we have separate stats collection methods.
-        let (key, key_stats) = codec_to_schema2::<K>(schemas.key.as_ref(), records.keys())
-            .map_err(|e| e.to_string())?;
-        let val = match updates {
-            BlobTraceUpdates::Row(_) => {
-                let (val, _) = codec_to_schema2::<V>(schemas.val.as_ref(), records.vals())
-                    .map_err(|e| e.to_string())?;
-                val
-            }
-            BlobTraceUpdates::Both(_, ext) => Arc::clone(&ext.val),
+        let ext = match updates.structured() {
+            Some(ext) => ext.clone(),
+            None => ColumnarRecordsStructuredExt {
+                key: codec_to_schema2::<K>(schemas.key.as_ref(), records.keys())
+                    .map_err(|e| e.to_string())?,
+                val: codec_to_schema2::<V>(schemas.val.as_ref(), records.vals())
+                    .map_err(|e| e.to_string())?,
+            },
         };
+
+        let key_stats = schemas
+            .key
+            .decoder_any(ext.key.as_ref())
+            .map_err(|e| e.to_string())?
+            .stats();
```

Comment:
This is the "important bit" - we no longer need to re-encode our input to collect stats. (Though in the short term we will continue to re-encode it for migration reasons!)